### PR TITLE
Crit switchleds

### PR DESCRIPTION
**Pull request by leahmariefernandez**
### Changelist 
- Switch LEDs are being read from VC
- Removed hardware function calls from the application layer and created a ShutdownSensor module
- Added tests for drive mode, shutdown sensor, LED status and CAN messages

### Testing Done
Unit tests and integration test. Did not test on board.

### CODE REVIEW

File: firmware/quadruna/CRIT/test/test_shutdownSensor.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+{
+    // Set the function to return a specific value when called with a specific sensor.
+    fake_io_shutdownSensor_readPin_returnsForArgs(&sensor, true);
+
+    // Call the function with the sensor.
+    io_shutdownSensor_readPin(&sensor);
+
+    // Verify that the function was called with the specific arguments.
+    ASSERT_EQ(1, fake_io_shutdownSensor_readPin_callCountForArgs(&sensor));
+}
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/test/test_shutdownSensor.cpp:48]: usually you want to assert on the app level effects of the code. This code is just testing if the fake io generation would work. For example, when a shutdown io pin is enabled, what happens? (usually the answer is something on CAN, or another peripheral, especially so for the CRIT).

File: firmware/quadruna/CRIT/src/app/app_mainState.c [hunk 2]
```diff
@@ -22,20 +23,24 @@ static void mainStateRunOnTick100Hz(void)
...
 
     const bool regen_switch_on = io_switch_isClosed(globals->config->regen_switch);
     app_canTx_CRIT_RegenSwitch_set(regen_switch_on ? SWITCH_ON : SWITCH_OFF);
-    io_led_enable(globals->config->regen_led, regen_switch_on);
 
     const bool torquevec_switch_on = io_switch_isClosed(globals->config->torquevec_switch);
     app_canTx_CRIT_TorqueVecSwitch_set(torquevec_switch_on ? SWITCH_ON : SWITCH_OFF);
-    io_led_enable(globals->config->torquevec_led, torquevec_switch_on);
+
+    if (app_canRx_VC_State_get() == 2)
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_mainState.c:38]: use the enum to compare

File: firmware/quadruna/CRIT/src/app/app_mainState.c [hunk 3]
```diff
@@ -22,20 +23,24 @@ static void mainStateRunOnTick100Hz(void)
...
+    {
+        io_led_enable(globals->config->start_led, true);
+        io_led_enable(globals->config->regen_led, true);
+        io_led_enable(globals->config->torquevec_led, true);
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_mainState.c:42]: idk about the logic here, i don't think this is correct. state drives the start led but not the regen/torque vectoring LEDs.

File: firmware/quadruna/CRIT/test/test_driveMode.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    // Verifying that the drive mode value is consistent across multiple reads
+    uint16_t consistentValue = 0b1010;
+    fake_io_driveMode_readPins_returns(consistentValue);
+
+    ASSERT_EQ(consistentValue, io_driveMode_readPins());
+    ASSERT_EQ(consistentValue, io_driveMode_readPins());
+
+    // The drive mode should be read consistently, and the fake should be called each time
+    ASSERT_EQ(2u, fake_io_driveMode_readPins_callCount());
+}
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/test/test_driveMode.cpp:49]: same thing here. Consider what the **last** effect of triggering the IO pin is.

File: firmware/quadruna/CRIT/test/test_ledStatus.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    ASSERT_EQ(1, fake_io_rgbLed_enable_callCountForArgs(&crit_status_led, false, false, true));
+}
+
+TEST_F(LedStatusTest, crit_board_status_led_control_with_no_error)
+{
+    // Don't set any error and check that the CRIT LED turns green
+    LetTimePass(10);
+    ASSERT_EQ(1, fake_io_rgbLed_enable_callCountForArgs(&crit_status_led, false, true, false));
+}
+
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/test/test_ledStatus.cpp:75]: CRIT test for LED turning red?

File: firmware/quadruna/CRIT/test/test_ledStatus.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    app_canRx_BMS_ImdOk_update(true);
+    app_canRx_BMS_BspdOk_update(true);
+    app_canRx_BMS_BmsOk_update(true);
+
+    // Set any critical error and check that the BMS LED turns red
+    app_canRx_BMS_Fault_CellOvertemp_update(true);
+    LetTimePass(10);
+
+    ASSERT_EQ(1, fake_io_rgbLed_enable_callCountForArgs(&bms_status_led, true, false, false));
+}
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/test/test_ledStatus.cpp:159]: test all the LEDs too, very good for regression and very important for safety.
SHDN especially.

possibly a very good idea to write a generic test, and pass in function pointers for leds and canerrors as arguments.

very strange that only VC/FSM/2 of CRIT is tested only

make sure that test ordering is also organized and logical.

**Review by Lucien950 — CHANGES_REQUESTED** [review]
solid work overall, just a few test things to fix. this is almost ready to drive.

File: firmware/quadruna/CRIT/src/app/app_mainState.c (hunk 3 above)
Comment (leahmariefernandez) [reply firmware/quadruna/CRIT/src/app/app_mainState.c]: ah ok i see. ok i moved the regen/torque vec LEDs to being read from switches.

File: firmware/quadruna/CRIT/src/io/io_driveMode.h [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 #include <stdbool.h>
 #include "app_utils.h"
+#include <stdint.h>
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/io/io_driveMode.h:5]:
```suggestion
```

**Review by Lucien950 — APPROVED** [review]

**Pull request merged**